### ISSUE TITLE
Ragged Batching for Image Data

### ISSUE THREAD

**Issue by langong347** [triton-inference-server/server#6984]
**Description**
1. I want to use dynamic batching for image requests of varying sizes (example below). Amongst the `BatchInput Kind` options, is [BATCH_ITEM_SHAPE](https://github.com/triton-inference-server/common/blob/00b3a71519e32e3bc954e9f0d067e155ef8f1a6c/protobuf/model_config.proto#L525) the correct option to use? Based on the docstring:

```
Among the requests in the batch, the shape of the
'source_input' will be added as input with shape
[batch_size, len(input_dim)]. For example, if one
batch-2 input with shape [3, 1] and batch-1 input
with shape [2, 2] are batched, the batch input will
have shape [3, 2] and value [ [3, 1], [3, 1], [2, 2]].
```

Example requests:
```
request A: [2, 3, 800, 600]  # (batch_size, channel, height, width)
request B: [1, 3, 795, 595]
```

2. Does the ONNX backend process ragged inputs and uniform (non-ragged) inputs in the same way? Based on the [Ragged Batching documentation page](https://docs.nvidia.com/deeplearning/triton-inference-server/user-guide/docs/user_guide/ragged_batching.html#ragged-batching) of how the backends handle ragged input, I'm concerned that processing the ragged inputs as a one-dimensional concatenated tensor will increase latency if uniform inputs can be processed in parallel:

_How ragged input are processed in a batch of requests depends on the backend implementation. The backends, such as [ONNX Runtime backend](https://github.com/triton-inference-server/onnxruntime_backend), [TensorFlow backend](https://github.com/triton-inference-server/tensorflow_backend), [PyTorch backend](https://github.com/triton-inference-server/pytorch_backend), and [TensorRT backend](https://github.com/triton-inference-server/tensorrt_backend), require models to accept ragged inputs as 1-dimensional tensors. These backends concatenates the request inputs into the 1-dimensional tensor._

3. Do I need to modify my model's API to add the `batch_input` as an additional argument, and then reshape the concatenated 1-dim tensor and pad them inside my model (as my model can only process data in array/tensor format)? So in order to use ragged batching, there'll be model change involved. Correct?

**Triton Information**
23.12

Are you using the Triton container or did you build it yourself?
Pre-built container from KServe

**To Reproduce**
```
import torch

req_a = torch.rand(2, 3, 800, 600)
req_b = torch.rand(1, 3, 795, 595)
```

The service runs the YOLOv8 model in ONNX format accepting tensors like above.

Describe the models (framework, inputs, outputs), ideally include the model configuration file (if using an ensemble include the model configuration file for that as well).

- YOLOv8 Model written in PyTorch
- ONNX Backend
- config.pbtxt
```
name: "my-yolov8-service"
platform: "onnxruntime_onnx"
max_batch_size: 8
input [
  {
    name: "images"
    data_type: TYPE_FP32
    dims: [ 3, -1, -1 ]
    allow_ragged_batch: true
  }
]
output [
  {
    name: "output0"
    data_type: TYPE_FP32
    dims: [ -1, -1 ]
  }
]
dynamic_batching {
    max_queue_delay_microseconds: 20000
}
batch_input [
  {
    kind: BATCH_ITEM_SHAPE
    target_name: "INDEX"
    data_type: TYPE_FP32
    source_input: "images"
  }
]
instance_group [
  {
    count: 2
    kind: KIND_GPU
  }
]
```

**Expected behavior**
1. For the example requests, using `BATCH_ITEM_SHAPE` will result in a batch input of shape `[3, 3]` and value `[[3, 800, 600], [3, 800, 600], [3, 795, 595]]`.

2. Performance (latency/throughput) should be similar with/without ragged batching. By without, I mean to minimally pad the requests to be of the same size.

3. I expect no model change, including API and internal logic, is required by using ragged batching. (but it seems like the main purpose of ragged batching is to allow requests of uneven lengths to be sent to Triton, but it will incur additional change to the model to reshape, and pad the uneven tensors.)

Linking similar issue #3951 
cc @GuanLuo
